### ISSUE TITLE
Clean Up NodeRun Lifecycle

### ISSUE THREAD

**Issue by diablouma** [littlehorse-enterprises/littlehorse#503]
{{NodeRunModel#advanceIfPossible()}} → Look at {{NodeRunModel#isCompleteOrRecoveredFromFailure()}}

Currently, we check for status {{EXCEPTION}} {{ERROR}} or {{HALTED}}. In the past, we only checked for {{ERROR}} (not even {{EXCEPTION}}).

A recent change (previous to PR 119, I believe it was one of the Wait For Threads PR’s) we made failures cause the {{ThreadRun}} to HALT, so the PR number 119 discovered a bug where the failure handlers for {{TASK}} node’s weren’t really working.

Now the {{if()}} statement checks for all three statuses ({{EXCEPTION}}, {{ERROR}}, {{HALTED}}), but I think that we only need to check for {{HALTED}}.

However, I am also concerned whether it is safe to check for the {{HALTED}} status, because {{HALTED}} means more things than “there was a failure handler”.

It might still be safe because in the {{ThreadRunModel#advance()}}, the ThreadRun itself still checks to ensure that its own status is {{RUNNING}} before calling the {{NodeRunModel#advanceIfPossible()}}

**Comment by coltmcnealy-lh** [comment]
@eduwercamacaro You and I should take a look at this in pair programming. This seems more tech debt than feature/bug.

**Comment by coltmcnealy-lh** [comment]
I tried to make a stylistic change to the `NodeRunModel#isCompleteOrRecoveredFromFailure()` to make it clearer in my first commit to #665. That change broke the e2e tests. Upon further discussion, we might actually want to refactor the `message NodeRun` protobuf. I am proposing:

1. Remove the `repeated int32 failure_handler_ids` field
2. Add `optional int32 failure_handler_id` to the `Failure` protobuf.

This would be a **BREAKING CHANGE** for currently active `WfRun`'s. However, we have confirmed with our early users that they are not using features that would be affected by this change. Additionally, we are still pre-1.0.0, so we have permission to make this change.
